player/lib/player/impl.ex

```diff
@@ -0,0 +1,66 @@
+defmodule Player.Impl do
+
+  defp shoot_if_shooting(player = %{shoot: true}) do
+    Bullet.new(player)
+    Map.put(player, :shoot, false)
+  end
+
+  defp shoot_if_shooting(player) do
+    player
+  end
+
+  def tick(player) do
+    player |>
+    Move.move |>
+    shoot_if_shooting
+  end
+
+  def action(player, :shoot) do
+    Map.put(player, :shoot, true)
+  end
+
+  def action(player, :stop) do
+    Map.put(player, :moving, false)
+  end
+
+  def action(player, direction) do
+    player |>
+    Map.put(:moving, true) |>
+    Map.put(:direction, direction)
+  end
+
+end
+
+defimpl Move, for: Player do
```

Comment:
Does valid_mode & move look like something that can be generalized to the Move protocol or should I keep this specific?

This would make "apply_move" to be the only required function to be implemented for something to use move assuming that 'valid_move' can be shared.  (Valid move will check boundaries & walls)

Reply:
I think this would make the Move protocol easier for me to reuse but it would sacrifice flexibility